docs: Add containerd install guide

jodh-intel opened the pull request:
Create a containerd installation guide.

Fixes: #738.

Signed-off-by: James O. D. Hunt <james.o.hunt@intel.com>

jodh-intel commented:
Distressingly, this only works for Debian/Ubuntu currently:

| Distro | Problem |
|-|-|
| CentOS / RHEL | No pre-built version of containerd for version 8? |
| Fedora | Fails due to cgroups v2 |
| OpenSUSE 15.2 | `containerd` package doesn't include service file or `ctr` command |

Can anyone comment / help with these issues?

jodh-intel commented:
@jcvenegas, @GabyCT - ptal.

liubin commented on docs/install/containerd/containerd-install-guide.md:
````diff
@@ -0,0 +1,99 @@
+   ```bash
+   $ tmpdir=$(mktemp -d)
+   $ pushd "$tmpdir" &>/dev/null
+   $ url="https://api.github.com/repos/kata-containers/runtime/releases/latest"
````
Shouldn't repo name be `kata-containers/kata-containers` ?

jodh-intel replied:
Not here - if you look carefully, this is the API URL rather than the "human readable" URL. This one returns JSON rather than XHTML ;)

jodh-intel replied:
Oh - sorry, I misunderstood what you were referring to. Yes we currently produce releases at https://github.com/kata-containers/runtime/releases, but I think we need to start using https://github.com/kata-containers/kata-containers/releases for 2.x releases. I've raised https://github.com/kata-containers/kata-containers/issues/745 to discuss this.

Note that the decision will affect https://github.com/kata-containers/kata-containers/pull/735, which needs to know where to look for releases.

liubin commented on docs/install/containerd/containerd-install-guide.md:
````diff
@@ -0,0 +1,99 @@
+
+   ```bash
+   $ sudo grep -q "io.containerd.kata.v2" "$cfg" || cat <<EOT | sudo tee -a "$cfg"
+   $ [plugins]
````
`$`s in config content should be removed?

jodh-intel replied:
Although the script that converts the doc to a script copes with this, I agree - this looks incorrect as shell doesn't show the same prompt for here docs. Fixed.

jodh-intel commented:
Updated to remove Debian as it doesn't provide a packaged version of containerd either. This means the guide only works for Ubuntu, which is a bit sad :sob:

jodh-intel commented:
/cc @amshinde.

bergwolf commented on docs/install/containerd/containerd-install.md:
````diff
@@ -0,0 +1,99 @@
+   ```bash
+   $ tmpdir=$(mktemp -d)
+   $ pushd "$tmpdir" &>/dev/null
+   $ url="https://api.github.com/repos/kata-containers/runtime/releases/latest"
````
The 2.0 releases are hosted here https://github.com/kata-containers/kata-containers/releases

bpradipt replied:
Since currently only pre-release version is available, the release url didn't work for me.  However I could verify rest of the instructions on Fedora 33
```
# kata-runtime --version
kata-runtime  : 2.0.0-rc0
   commit   : 84af1c20f34f9288bde422dc8bf7a0627b81c992
   OCI specs: 1.0.1-dev
# ctr run --runtime "io.containerd.kata.v2" --rm -t "$image" test-kata uname -m

x86_64
```

jodh-intel replied:
Thanks for testing @bpradipt !

jodh-intel replied:
@jcvenegas, @amshinde, @gnawux, @bergwolf - We have a bit of a problem here. Since [as @jcvenegas pointed out the `latest` URL is not guaranteed to list the latest release](https://github.com/kata-containers/kata-containers/pull/735#pullrequestreview-490811033), what do we do? For Kata 1.x, we rely on the distros package manager to determine the latest version, but here we can't.

Ironically, we *do* now have the facility to determine the latest release thanks to the `kata-check` changes on https://github.com/kata-containers/kata-containers/pull/735. But we then have a "chicken and egg" scenario - the user needs to know the latest version to download, but can't determine that without downloading a version including that change to `kata-check` :sob:

Maybe we could create a very small YAML file listing the latest releases for 1.x, 2.x and stable releases and host it on https://katacontainers.io (a bit like a cross between https://download.clearlinux.org/latest and the kernel version summary on https://www.kernel.org). That would require changes to our release tooling, `kata-check` and this document, but would solve the problem and would also isolate us from this odd GitHub behaviour.

jodh-intel replied:
I think I've found a solution:

```bash
$ url="https://api.github.com/repos/kata-containers/kata-containers/releases"
$ latest_release=$(curl -sL "$url" | jq -r '.[].tag_name | select(contains("-") | not)' | sort -t "." -k1,1n -k2,2n -k3,3n | tail -1)
$ echo $latest_release
1.11.2
```

jodh-intel replied:
Hmm. That's the result using the 2.x URL (which isn't quite right). But with 1.x:

```bash
$ url="https://api.github.com/repos/kata-containers/runtime/releases"
$ curl -sL "$url" | jq -r '.[].tag_name | select(contains("-") | not)' | sort -t "." -k1,1n -k2,2n -k3,3n | tail -1
1.11.3
```

jodh-intel replied:
Eek - actually, that 1.11.2 *is* correct. tal at:

- https://github.com/kata-containers/runtime/releases/tag/1.11.3

  Looks fine.

- https://github.com/kata-containers/kata-containers/releases/tag/untagged-50bda11e5dc23f08b6ce

  Huh? That is the URL you get when you click "Kata Containers 1.11.3". It also hows as "draft" and is missing all assets!?!

/cc @egernst, @bergwolf.

bergwolf replied:
@jodh-intel I think we need to determine which release we want to download for 2.0 here. Right now there is no stable 2.0 release so all the releases are pre-release. Maybe we can just point users to `https://github.com/kata-containers/kata-containers/blob/2.0-dev/VERSION` to pick up the latest version. And once we tag the first 2.0 stable release, we can move to use `https://api.github.com/repos/kata-containers/kata-containers/releases/latest`.

In the meantime, we should stop tagging `1.x` releases on `kata-containers` repository. There is practically nothing for `1.x` to release here and we are using the runtime repository for the `1.x` releases.

WDYT?

jodh-intel replied:
> Maybe we can just point users to https://github.com/kata-containers/kata-containers/blob/2.0-dev/VERSION

That's a good idea. But, assuming we merge `2.0-dev` into `master` when we release 2.0 (and I was assuming we'd do that), this idea would break the installation guide. However, why can't we merge `2.0-dev` into `master` "now" since this repo is purely for 2.x development? That way the doc can safely refer to https://github.com/kata-containers/kata-containers/blob/master/VERSION as that URL will then become "reliable".

> In the meantime, we should stop tagging 1.x releases on kata-containers repository.

Agreed. It's confusing to me (and maybe others) that we show 1.x releases in the 2.x repo.

/cc @amshinde, @egernst, @jcvenegas.

jodh-intel replied:
Regarding the doc referring to https://github.com/kata-containers/kata-containers/blob/2.0-dev/VERSION, there *is* another solution to avoid breaking the doc - make the doc query the default branch for this repo. However, that's an extra step which arguably the install doc shouldn't need to do.

bergwolf replied:
We already have several places referencing `2.0-dev` branch of `kata-containers`. How about just using it for now and fixing all these places once we decide to move from `2.0-dev` to `master`.

devimc commented on docs/install/containerd/containerd-install.md:
```diff
@@ -0,0 +1,99 @@
+   $ source /etc/os-release || source /usr/lib/os-release
+   $ packages='containerd curl git jq'
+   $ case "$ID" in
+   $   ubuntu) sudo apt-get -y install $packages ;;
```
debian should work too, shouldn't it?

jodh-intel replied:
In theory, but see https://github.com/kata-containers/kata-containers/pull/739#issuecomment-693443741.

bpradipt replied:
These packages including `containerd` are available in Fedora as well.. Can we include `fedora` case as well ? 
cc @fidencio

jodh-intel replied:
@bpradipt - I'd like to, but see https://github.com/kata-containers/kata-containers/pull/739#issuecomment-693443741.

bpradipt replied:
Ah I see. Yeah that's a problem. Can we add a comment that Fedora needs to be booted with cgroup v1 ? (`systemd.unified_cgroup_hierarchy=0`)

jodh-intel replied:
Yes. It's a shame as it means we cannot "test" the document in the CI on Fedora 32. I plan to change the doc to error if cgroups v2 is detected, along with a helpful message.

I've discussed the general approach with @chavafg and as a result of that currently reworking this doc to install containerd from the containerd binary packages rather than distro packaged (due to all the current problems related to the containerd distro packages). That will allow the doc to work for all distros (except F32 - see above). I'll also add a warning since the user will need to consider updating Kata and containerd as and when new releases are available.

devimc commented on docs/install/containerd/containerd-install.md:
```diff
@@ -0,0 +1,99 @@
+     			runtime_type = "io.containerd.kata.v2"
+     		[plugins.linux]
+     			shim = "/usr/bin/containerd-shim"
+     			shim_debug = true
```
uhmm last time I tried to enable shim debug with this option, it didn't work, I add to add the following

```
[debug]
  level = "debug"
[plugins]
     	[plugins.cri]
..........
```

jodh-intel replied:
Thanks - I'll double-check that tomorrow...

devimc replied:
I can confirm that this doesn't work for me, I need to add `[debug] level="debug"`

jodh-intel replied:
@devimc - I don't get debug even with that :( What version of containerd are you using (I'm trying with v1.4.1).

devimc replied:
I'm using `containerd v1.3.2`. This is my config file

```toml
[debug]
  level = "debug"
[plugins]
  [plugins.cri]
    [plugins.cri.containerd]
      [plugins.cri.containerd.runtimes]
        [plugins.cri.containerd.runtimes.runc]
           runtime_type = "io.containerd.runc.v1"
           [plugins.cri.containerd.runtimes.runc.options]
             BinaryName = "/usr/bin/runc"
             Root = ""
        [plugins.cri.containerd.runtimes.kata]
           runtime_type = "io.containerd.kata.v2"
           privileged_without_host_devices = true
           pod_annotations = ["io.katacontainers.*"]
        [plugins.linux]
           shim = "/usr/bin/containerd-shim"
           shim_debug = true
```

eadamsintel replied:
I happened to see this discussion in my email.  It would be easier for me if the config showed how to enable for all of the supported hypervisors. At the very least the instructions should show how to enable annotations like the example above shows.  Isn't `kata-deploy` becoming the recommended installation method?  There is a kata 2.0 rc0 on Docker hub for `kata-deploy`. To make this more generic, I would consider using that installation method.

devimc reviewed (approved):
thanks @jodh-intel - two nits that can be ignored

fidencio commented:
@jodh-intel, for CentOS / RHEL 8, what about suggesting using: https://download.docker.com/linux/centos/docker-ce.repo

jodh-intel commented:
@fidencio - thanks. I've updated the table in https://github.com/kata-containers/kata-containers/pull/739#issuecomment-693443741.

fidencio commented on docs/install/containerd/containerd-install.md:
````diff
@@ -0,0 +1,181 @@
+   Check to see if Kata Containers is already installed:
+
+   ```bash
+   $ command -v kata-runtime &>/dev/null && die "Please remove existing Kata Containers installation"
````
Checking for kata-runtime will work well for 1.x branches, but if we want to make this bullet proof, checking for the containerd-shim-kata-v2 binary could be a better call.

jodh-intel replied:
Agreed - done.

fidencio commented on docs/install/containerd/containerd-install.md:
```diff
@@ -0,0 +1,181 @@
+   $ source /etc/os-release || source /usr/lib/os-release
+   $ packages='curl git jq'
+   $ case "$ID" in
+       centos|rhel) sudo yum -y remove $packages ;;
```
Please, replace remove by install.

jodh-intel replied:
Good catch! Updated.

fidencio commented on docs/install/containerd/containerd-install.md:
````diff
@@ -0,0 +1,181 @@
+
+   ```bash
+
+   $ url="https://api.github.com/repos/kata-containers/runtime/releases"
````
Do we still want to point to the runtime repo?

jodh-intel replied:
Fixed.

fidencio commented on docs/install/containerd/containerd-install.md:
````diff
@@ -0,0 +1,181 @@
+
+   ```bash
+   $ sudo ln -sf /opt/kata/bin/containerd-shim-kata-v2 /usr/bin/
+   $ sudo ln -sf /opt/kata/bin/kata-runtime /usr/bin
````
Another reference to the kata-runtime here.

jodh-intel replied:
Can you elaborate?

fidencio replied:
@jodh-intel, my bad, just ignore this one. :-)

fidencio reviewed (commented):
@jodh-intel, I've made a few comments in the review itself, please take a look.

In case I missed the decision about mentioning `kata-runtime` here and there, please, just "resolve" the thread.

jodh-intel commented:
I've updated the doc and tested on Debian. However, we need to decide if this approach is the right one.

# Summary

Pros:

+ Single document to handle all distros.
+ It works :smile:

Cons:

- Arguably too complex / long.
- Installs binaries outside of a package manager.

# Questions

- Should the code in this document be put into a stand-alone script instead? Or into [`kata-manager`](https://github.com/kata-containers/tests/tree/master/cmd/kata-manager)?

  **Pros:** Would simplify the doc (just `curl $url | sh`).
  **Cons:** Not good for users who want to undestand each step of the installation.
   
- Should we add this functionaltiy to [`kata-deploy`](https://github.com/kata-containers/kata-containers/tree/2.0-dev/tools/packaging/kata-deploy)?

  **Pros:** Avoids duplication.
  **Cons:**
  - `kata-deploy` is specifically for k8s cloud installations, not for user installs.
  - Would require lots of changes just to make it friendly for users.

- Should we split this document into separate per-distro docs?

  **Pros:**
  + Clearer logic (as docs won't need to cater for other distros).
  + "Safer" as some distros already provide `containerd` (which would therefore be under package manager control).
  + We could encourage users to use snap for Ubuntu.

  **Cons:** Potentially different approach for each distro.

This PR is release gating but imho we need to decide on the best approach. This PR is "done"/"ready", but could be considered "blocked" too until we have a consensus on the best approach. Hence, comments please folk! :smile:

/cc @kata-containers/architecture-committee, @kata-containers/documentation *et al*.

bergwolf commented:
IMHO, since we have stopped building/shipping distro packages, I think we should provide an upstream way of installation which is installing via a single document. For the distributions that ship distro packages, the distro package maintainers can add distro official guide whenever they deem necessary. Just my two cents.

jodh-intel commented:
@bergwolf - that's a good point and thanks for raising this! I think it will take some time for the distros to "catch up" and package a 2.0 release. On that topic, I've updated the distro packaging status table:

- https://github.com/kata-containers/packaging/wiki/Distro-packaging-status

fwics it is only Debian which doesn't package Kata.

# Kata 1.x releases

However, maybe we should update https://github.com/kata-containers/documentation/tree/master/install/ for those distros that now provide distro packages, suggesting users install those rather than the OBS packaged versions?

We could optionally retain the existing installation guides for those that want to use the latest releases.

# Kata 2.x releases

Summary:

- We have decided not to produce OBS packages for 2.x
- Distros will take a while to "catch up" and update their packages to 2.x.

So, what do we do when 2.0 launches?

1. 2.0 is released.
1. Source and binary packages are uploaded to https://github.com/kata-containers/kata-containers/releases.
1. What now?
   - Wait for the distros
     - "Do nothing" for a while as we wait for the distros to update their packaging.
     - Once that happens, create a set of distro-specific install guides, showing how to install Kata using the distro-packages.

   - Offer a binary installation guide (aka this PR).
     This is "similar" to the existing 1.x install guides in that the user will install the latest version, but inferior in that it won't ever be updated automatically.
   
1. When distros offer native 2.x packages, update the containerd install guide to error and asking users to install Kata using their distros package manager? Or maybe do that but retain this install guide for users who want the latest.


# TL;DR

I've updated the doc with more warnings and a note about using distro-packaged versions where available. It's purposely vague since we cannot (yet) point users at a distro-packaged versions of Kata 2.0 obviously.

jodh-intel commented:
I've just updated the doc again, simplifying the logic by leveraging `hub`! ;) But imho the current version of this document is still arguably too complex. Admittedly, if we update `kata-manager`, the user won't need to actually *run* the code manually, but still it's not ideal imho.

What do others think?

jodh-intel commented:
/test

fidencio reviewed (approved):
lgtm!

jodh-intel commented:
Marking as `dnm` since `hub` required `git` and `GITHUB_*` variables to be setup, but we cannot assume users will either have done that (or want to).

jodh-intel commented:
# New plan

Given the amount of shell code needed to install and configure both projects, plus the issues around using `hub`, on reflection, I think the simplest approach for this PR is:

- Have a doc that essentially tells users to:

  - "install containerd"

    Either the latest version (which potentially won't work), or the known good version from https://github.com/kata-containers/kata-containers/blob/2.0-dev/versions.yaml)

  - "install Kata". just provide the https://github.com/kata-containers/kata-containers/releases link and explain what users should do.

  - Provide config details as code in the doc.

- Create a script that users can view and or `curl $script | sh` if they want.

  I was thinking about this URL for the location:

  - https://github.com/kata-containers/kata-containers/tree/2.0-dev/docs/install/scripts/kata-install.sh.

    This script will be stand-alone and will contain all the shell code needed to install + configure both containerd + Kata 2.x.

- Create a periodic GitHub action (daily? weekly?) that runs the script on atleast Ubuntu to ensure Kata remains installable from the *script*.

# TL;DR

Simplify the doc, make it more generic and don't include big chunks of shell code in it. By doing this, we can avoid scaring off new users with a brief document but also retain the ease of installation we currently have with Kata 1.x.

Thoughts?

jodh-intel commented:
Ok - silence is compliance :smile:

I've totally refreshed this PR to adopt the new plan.

I've tested the new 2.0 `kata-manager.sh` script on Debian 10, Ubuntu 20.04 and CentOS 8. Fails on F32 due to cgroups v2 so if anyone has thoughts on how to deal with that, please let me know. ftr, I don't want the script to modify the system boot just to install Kata.

/cc @fidencio, @c3d.

bergwolf commented:
@jodh-intel Where did it fail? Does `kata-manager.sh` check host cgroups layout?

jodh-intel commented:
@bergwolf, @amshinde - Here's some more details on running on F32:

```bash
$ containerd-shim-kata-v2 --version
Kata Containers containerd shim: id: "io.containerd.kata.v2", version: 2.0.0-rc1, commit: 9b969bb7da4b37c25ad3a2512e74bb02129381f1

$ mount|grep cgroup
cgroup2 on /sys/fs/cgroup type cgroup2 (rw,nosuid,nodev,noexec,relatime,seclabel,nsdelegate)

$ sudo ctr --debug run --rm --runtime "io.containerd.kata.v2" -t docker.io/library/busybox:latest test-kata uname -m
ctr: Could not create cgroup for /default/kata_test-kata: cgroups: cgroup mountpoint does not exist: not found

$ sudo journalctl -t kata|grep level=error
oct. 09 17:00:09 f32-vm kata[3149]: time="2020-10-09T17:00:09.805383825+01:00" level=error msg="container create failed" error="Could not create cgroup for /default/kata_test-kata: cgroups: cgroup mountpoint does not exist" pid=3149 sandbox=test-kata source=virtcontainers subsystem=container
```

bergwolf commented:
@jodh-intel Does containerd + runc work on F32? Looks like the error comes from containerd not kata.

bergwolf commented:
Ok, I see. The problem is that kata runtime does not support host cgroups v2 yet. Then my suggestion is that we choose an older Fedora release (such as F30?) to support until kata has supported host cgroups v2.

Ref:
https://github.com/kata-containers/kata-containers/blob/2.0-dev/src/runtime/virtcontainers/sandbox.go#L1989
https://github.com/kata-containers/runtime/blob/master/virtcontainers/sandbox.go#L2000

jodh-intel commented:
Hi @bergwolf - that would be ideal, but we can't as F30 is EOL and the [cgroup problem started in F31](https://fedoraproject.org/wiki/Releases/31/ChangeSet#Modify_Fedora_31_to_use_CgroupsV2_by_default) ;( The best I can do is detect a distro that only mounts the cgroups v2 mount and exit immediately with a helpful error message.

fidencio commented:
> Hi @bergwolf - that would be ideal, but we can't as F30 is EOL and the [cgroup problem started in F31](https://fedoraproject.org/wiki/Releases/31/ChangeSet#Modify_Fedora_31_to_use_CgroupsV2_by_default) ;( The best I can do is detect a distro that only mounts the cgroups v2 mount and exit immediately with a helpful error message.

@jodh-intel, sorry, I totally missed the /cc a few days ago. Unfortunately the best we can do is follow your suggestion.

jodh-intel commented:
@bergwolf - yep - I've raised #927 to track that. That issue URL is also displayed by the script and specified in the docs on this PR.

Re-tested and the script now correctly exits on F32 but continues the install on other distro versions:

```sh
INFO: Running pre-checks
ERROR: Kata Containers does not yet fully support cgroups v2 - see https://github.com/kata-containers/kata-containers/issues/927
```

It would be great if someone else could validate the script runs successfully for them on a test system.. :smile:

jodh-intel commented:
/test

jodh-intel commented:
/test

bergwolf commented:
`debian-ci` failed on `not ok 1 Verify nginx connectivity between pods`. IIRC it was fixed by updating `flannel` version. Maybe you need to rebase?

jodh-intel commented:
Thanks @bergwolf.

In case folks still want to test this prior to the official 2.0 release, here's how:

```bash
$ version="2.0.0-rc1"
$ bash -c "$(curl -fsSL https://raw.githubusercontent.com/kata-containers/kata-containers/2.0-dev/utils/kata-manager.sh) $version"
```

Obviously, you can update `$version` as new releases appear on https://github.com/kata-containers/kata-containers/releases. If you don't specify a version, you'll get the latest official release. But note **that will be a 1.x release until 2.0 is released!!**
